packages/edit-widgets/src/components/widget-areas-block-editor-provider/index.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * External dependencies
+ */
+import { defaultTo } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	DropZoneProvider,
+	SlotFillProvider,
+	FocusReturnProvider,
+} from '@wordpress/components';
+import { uploadMedia } from '@wordpress/media-utils';
+import { useSelect } from '@wordpress/data';
+import { useEffect, useMemo, useState } from '@wordpress/element';
+import { createBlock } from '@wordpress/blocks';
+import {
+	BlockEditorProvider,
+	BlockEditorKeyboardShortcuts,
+} from '@wordpress/block-editor';
+
+/**
+ * Internal dependencies
+ */
+import KeyboardShortcuts from '../keyboard-shortcuts';
+
+const EMPTY_ARRAY = [];
+
+export default function WidgetAreasBlockEditorProvider( {
+	blockEditorSettings,
+	...props
+} ) {
+	const { areas, hasUploadPermissions } = useSelect( ( select ) => {
+		const { canUser, getEntityRecords } = select( 'core' );
+		return {
+			areas: getEntityRecords( 'root', 'widgetArea' ) || EMPTY_ARRAY,
+			hasUploadPermissions: defaultTo(
+				canUser( 'create', 'media' ),
+				true
+			),
+		};
+	} );
+	const [ blocks, setBlocks ] = useState( [] );
+	useEffect( () => {
+		if ( ! areas || blocks.length > 0 ) {
+			return;
+		}
+		setBlocks(
+			areas.map( ( { id, name }, index ) => {
+				return createBlock( 'core/widget-area', {
+					id,
+					name,
+					isOpen: index === 0,
+				} );
+			} )
+		);
+	}, [ areas ] );
```

Comment:
Should `blocks` also be listed in the dependency array?

Reply:
Nice catch @ZebulanStanphill, I think we should list it 👍